### DISCUSSION TITLE
Split SBT manager

### DISCUSSION THREAD

**Discussion by gaeljw** [renovatebot/renovate#20676] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

Following https://github.com/renovatebot/renovate/issues/20481, the SBT manager now support generating PR for SBT version itself.

It works great until you're in some enterprise context with private registry.

To give some context, the SBT manager handles two kind of dependencies:
- regular SBT/Maven dependencies (Java/Scala libraries typically), Maven datasource, with default registry being Maven central
- SBT itself, with GitHub releases datasource, with default registry GitHub.com

Here's the scenario I encountered:
- already using Renovate in SBT projects (before it supported SBT version upgrade)
- having a custom Renovate configuration with private registry for SBT/Maven dependencies. Something like this:
```json
{
  "matchManagers": ["sbt"],
  "registryUrls": [
      "https://artifactory.mycompany.net/artifactory/xxx"
   ]
}
```
- after upgrade to new Renovate version, I was not getting SBT version upgrades, so I had to enable logs to find out that SBT version upgrades were looked for in the private registry rather than on GitHub releases.
- I then added GitHub.com in the registry URLs in config above but I got following error: `WARN: Excess registryUrls found for datasource lookup - using first configured only
datasource": "github-releases",`
- finally, I added GitHub.com as the first registry URL in config above and it worked:
```json
{
  "matchManagers": ["sbt"],
  "registryUrls": [
      "https://github.com",
      "https://artifactory.mycompany.net/artifactory/xxx"
   ]
}
```

IMHO, there are several issues here:
- if I hadn't worked on the PR myself, I would have never thought I needed to add GitHub.com in the registries: why would I need GitHub.com for? I should not have to know the implementation details that SBT versions upgrades are checked against GitHub releases.
- only SBT version upgrade need to be checked against GitHub.com but all other regular libraries upgrades have to be checked against the (private) Maven registry. With my last configuration above, it means all libs upgrades will be searched for on GitHub first and only in 2nd step in the private registry. This is not optimal.

I'm wondering if we should not clearly split the SBT manager in two pieces: regular libs manager, and sbt itself manager. Keeping the "SBT" name for the regular libs manager and creating a new manager for SBT itself.

**Comment by gaeljw** [comment]
And now that I wrote all of this, I just checked and it appears that SBT is also published as a JAR/Maven dependency.

**Thus the easy way around this would actually be to check SBT upgrades with Maven datasources as for regular libs.**

See https://mvnrepository.com/artifact/org.scala-sbt/sbt/1.8.2, the artifact `org.scala-sbt.sbt` could be used.

Only downside would be loosing the changelog I guess because I don't see a Changelog.md in the repo referenced by the JAR. (Not sure of all the rules that apply to look for a changelog 🤷‍♂️).

I'll be happy to rework on this myself.

**Reply by rarkins** [reply]
A manual changelog mapping could be added

**Reply by gaeljw** [reply]
Cool, didn't know that.

**Comment by gaeljw** [comment]
Behind this issue that will likely be solved by using only the Maven datasource, I'm wondering if the custom configuration of registries should not be possible by datasource rather than by manager. Different managers using same datasource are likely to need same custom registries if any.

**Reply by viceice** [reply]
that's already possible, just use `matchDatasources` instead of `matchManagers`.

we already prefer that when suggesting help

**Reply by gaeljw** [reply]
Good to know, will use this then.
